src/js/contentAwareRouter.js

```diff
@@ -0,0 +1,107 @@
+//   }
+// }
+//
+// The `Accepts` headers supplied by the user will be tested using `request.accepts(contentType)`.  The first
+// `contentType` that matches will be used.  If no `Accepts` headers are provided, any contentType will match, and
+// hence the first handler will be used.   Thus, you should add handlers in the order you would prefer that they are
+// used.
+//
+// When the decision has been made and a match has been found, a dynamic component will be constructed using the
+// appropriate 'handlerGrades'
+//
+"use strict";
+var fluid = fluid || require("infusion");
+var gpii  = fluid.registerNamespace("gpii");
+
+require("./requestAwareRouter");
+
+fluid.registerNamespace("gpii.express.contentAware.router");
+gpii.express.contentAware.router.delegateToHandler = function (that, request, response) {
+    var handlerGrades = gpii.express.contentAware.router.getHandlerGradesByContentType(that, request, that.options.handlers);
+    if (handlerGrades) {
+        // Add a third argument with the list of grades.
+        that.events.onRequest.fire(request, response, handlerGrades);
+    }
+    else {
+        response.status(500).send({ok: false, message: "Could not find an appropriate handler for the content types you accept."});
+    }
+};
+
+gpii.express.contentAware.router.getHandlerGradesByContentType = function (that, request, handlers) {
+    var handlerGrades = null;
+    fluid.each(handlers, function (value) {
+        if (!handlerGrades && Boolean(request.accepts(value.contentType))) {
+            handlerGrades = value.handlerGrades;
+        }
+    });
+
+    return handlerGrades;
+};
+
+gpii.express.contentAware.router.getHandler = function (that) {
+    return that.delegateToHandler;
+};
+
+fluid.defaults("gpii.express.contentAware.router", {
+    gradeNames: ["gpii.express.router"],
+    events: {
+        onRequest: null
+    },
+    dynamicComponents: {
+        // This intermediate structure is required because we cannot directly refer to the original list of grades
+        // from a `gradeNames` option.  This will be resolved once we upgrade beyond the versions of infusion affected by:
+        // https://issues.fluidproject.org/browse/FLUID-5742
+        broker: {
+            type:          "fluid.component",
+            createOnEvent: "onRequest",
+            options: {
+                gradeNames:    ["gpii.express.contentAware.broker"],
+                mergePolicy: {
+                    "request":  "nomerge",
+                    "response": "nomerge"
+                },
+                request:       "{arguments}.0",
+                response:      "{arguments}.1",
+                handlerGrades: "{arguments}.2",
+                components: {
+                    requestHandler: {
+                        type:          "gpii.express.handler",
+                        options: {
+                            timeout:    "{router}.options.timeout",
```

Comment:
This is awkward and will end up trying to source this option from anything named "router", including your immediate parent. I'd suggest removing this argument entirely and for the time being tell your users to use a distributeOptions block in the rare case they want to configure this parameter. In the long term, it would be cleaner to turn your timeout handler into a grade of its own which can then be independently configured stably. If you want to keep this argument here, you should increase the scope on this argument to match more exactly a context you expect it to be found in. Note that you do in fact configure this option inside gpii.express.handler, rather than in a router. Neither is a great place for it, but it should at least be clear to a reader where the standard location is from which they can override it. Given this behaviour is currently hardwired into the request handling machinery itself, I suggest that it be sourced from something in the gpii.express config itself at top level.